File: IO.Astrodynamics.Net/IO.Astrodynamics/Maneuver/Launch.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using IO.Astrodynamics.Frames;
using IO.Astrodynamics.Math;
using IO.Astrodynamics.Surface;
using IO.Astrodynamics.TimeSystem;

namespace IO.Astrodynamics.Maneuver
{
    public class Launch
    {
        private double _nonInertialAscendingAzimuthLaunch = double.NaN;
        private double _nonInertialDescendingAzimuthLaunch = double.NaN;
        private double _inertialAscendingAzimuthLaunch = double.NaN;
        private double _inertialDescendingAzimuthLaunch = double.NaN;
        private double _nonInertialInsertionVelocity = double.NaN;
        private readonly double _deltaL = 0.0;

        public LaunchSite LaunchSite { get; }
        public Site RecoverySite { get; }
        public OrbitalParameters.OrbitalParameters TargetOrbit { get; }
        public bool? LaunchByDay { get; }
        public double Twilight { get; }


        /// <summary>
        /// 
        /// </summary>
        /// <param name="launchSite">Launch from</param>
        /// <param name="recoverySite">Recovery site</param>
        /// <param name="targetOrbit">Orbit to reach</param>
        /// <param name="twilight"></param>
        /// <param name="launchByDay">Define if launch should occur by day. If undefined launch can occur everytime</param>
        /// <exception cref="ArgumentNullException"></exception>
        public Launch(LaunchSite launchSite, Site recoverySite, OrbitalParameters.OrbitalParameters targetOrbit,
            double twilight, bool? launchByDay)
        {
            LaunchSite = launchSite ?? throw new ArgumentNullException(nameof(launchSite));
            RecoverySite = recoverySite ?? throw new ArgumentNullException(nameof(recoverySite));
            TargetOrbit = targetOrbit ?? throw new ArgumentNullException(nameof(targetOrbit));
            LaunchByDay = launchByDay;
            Twilight = twilight;
            _deltaL= System.Math.Asin(System.Math.Tan(LaunchSite.Planetodetic.Latitude) / System.Math.Tan(targetOrbit.Inclination()));
        }

        /// <summary>
        /// Find launch windows based on launch's constraints in the given window
        /// </summary>
        /// <param name="searchWindow"></param>
        /// <param name="outputDirectory"></param>
        /// <returns></returns>
        public IEnumerable<LaunchWindow> FindLaunchWindows(in Window searchWindow, DirectoryInfo outputDirectory)
        {
            //return API.Instance.FindLaunchWindows(this, searchWindow, outputDirectory);
            var date = searchWindow.StartDate;

            // Define initial step (half the length of the search window)
            TimeSpan step = searchWindow.Length * 0.5;

            // Define crossing plane inbound status
            bool isInboundPlaneCrossing = IsInboundPlaneCrossing(date);

            List<LaunchWindow> launchWindows = new List<LaunchWindow>();

            // Work until it reaches the end of the search window
            while (date <= searchWindow.EndDate)
            {
                date = date.Add(step);

                // Compute crossing plane inbound status
                bool isInbound = IsInboundPlaneCrossing(date);

                // If inbound status has changed, we passed through the orbital plane
                if (isInbound != isInboundPlaneCrossing)
                {
                    isInboundPlaneCrossing = isInbound;
                    // Reduce step size and reverse search direction
                    step = TimeSpan.FromTicks(-step.Ticks / 2);
                }

                // If sufficient accuracy is reached (step size less than 1 second)
                if (System.Math.Abs(step.TotalSeconds) < 1.0)
                {
                    // Fill in the first launch window

                    double inertialAzimuthLaunch = double.NaN;
                    double nonInertialAzimuthLaunch = double.NaN;

                    // Set velocities
                    double inertialInsertionVelocity = GetInertialInsertionVelocity();
                    double nonInertialInsertionVelocity = GetNonInertialInsertionVelocity();

                    // Determine if launch will be ascending or descending and set azimuths
                    bool isAscending;
                    if (IsAscending(date))
                    {
                        inertialAzimuthLaunch = GetInertialAscendingAzimuthLaunch();
                        nonInertialAzimuthLaunch = GetNonInertialAscendingAzimuthLaunch();
                        isAscending = true;
                    }
                    else
                    {
                        inertialAzimuthLaunch = GetInertialDescendingAzimuthLaunch();
                        nonInertialAzimuthLaunch = GetNonInertialDescendingAzimuthLaunch();
                        isAscending = false;
                    }

                    // Add the first launch window to the collection
                    Window window = new Window(date, date);
                    launchWindows.Add(new LaunchWindow(window,  inertialInsertionVelocity,
                        nonInertialInsertionVelocity,inertialAzimuthLaunch, nonInertialAzimuthLaunch));

                    // Compute following launch windows
                    TimeSpan remainingTime = searchWindow.EndDate - date;

                    // Sidereal rotation period
                    TimeSpan siderealRotation = LaunchSite.CelestialBody.SideralRotationPeriod(date);
                    TimeSpan halfSiderealRotation = TimeSpan.FromTicks(siderealRotation.Ticks / 2);

                    // Number of remaining launch windows
                    int nbLaunchWindows = (int)(remainingTime.TotalSeconds / halfSiderealRotation.TotalSeconds);

                    for (int i = 0; i < nbLaunchWindows; i++)
                    {
                        // If previous is ascending, next will be descending
                        isAscending = !isAscending;

                        double deltaLDuration = (_deltaL * 2.0) / LaunchSite.CelestialBody.AngularVelocity(date);

                        if (!isAscending)
                        {
                            deltaLDuration *= -1.0;
                        }

                        // Increment next launch date
                        date = date.Add(halfSiderealRotation).Add(TimeSpan.FromSeconds(deltaLDuration));

                        // Build next window
                        Window nextWindow = new Window(date, date);

                        if (isAscending)
                        {
                            launchWindows.Add(new LaunchWindow(nextWindow,inertialInsertionVelocity,nonInertialInsertionVelocity, GetInertialAscendingAzimuthLaunch(), GetNonInertialAscendingAzimuthLaunch()));
                        }
                        else
                        {
                            launchWindows.Add(new LaunchWindow(nextWindow,inertialInsertionVelocity,nonInertialInsertionVelocity, GetInertialDescendingAzimuthLaunch(), GetNonInertialDescendingAzimuthLaunch()));
                        }
                    }

                    break; // Exit the while loop after computing all launch windows
                }
            }

            if (LaunchByDay != null)
            {
                var launchSiteDayWindows = LaunchSite.FindDayWindows(searchWindow, Twilight);
                var recoverySiteDayWindows = RecoverySite.FindDayWindows(searchWindow, Twilight);
                bool launchByDay = LaunchByDay.Value;

                foreach (var launchWindow in launchWindows.ToArray())
                {
                    bool launchSiteIntersects = launchSiteDayWindows.Any(x => x.Intersects(launchWindow.Window));
                    bool recoverySiteIntersects = recoverySiteDayWindows.Any(x => x.Intersects(launchWindow.Window));

                    bool removeWindow = launchByDay
                        ? !(launchSiteIntersects && recoverySiteIntersects)
                        : (launchSiteIntersects || recoverySiteIntersects);

                    if (removeWindow)
                    {
                        launchWindows.Remove(launchWindow);
                    }
                }
            }

            return launchWindows;
        }

        public double GetNonInertialAscendingAzimuthLaunch()
        {
            if (double.IsNaN(_nonInertialAscendingAzimuthLaunch))
            {
                double inertialInsertionVelocity = GetInertialInsertionVelocity();
                double inertialAscendingAzimuthLaunch = GetInertialAscendingAzimuthLaunch();

                double launchSiteVelocityMagnitude = LaunchSite.GetEphemeris(Time.J2000TDB, LaunchSite.CelestialBody,Frame.ICRF, Aberration.None).ToStateVector().Velocity.Magnitude();

                double vrotx = inertialInsertionVelocity * System.Math.Sin(inertialAscendingAzimuthLaunch) -
                               launchSiteVelocityMagnitude;
                double vroty = inertialInsertionVelocity * System.Math.Cos(inertialAscendingAzimuthLaunch);

                _nonInertialAscendingAzimuthLaunch = System.Math.Atan2(vrotx, vroty);
                if (_nonInertialAscendingAzimuthLaunch < 0.0)
                {
                    _nonInertialAscendingAzimuthLaunch += Constants._2PI;
                }
            }

            return _nonInertialAscendingAzimuthLaunch;
        }

        public double GetNonInertialDescendingAzimuthLaunch()
        {
            if (double.IsNaN(_nonInertialDescendingAzimuthLaunch))
            {
                _nonInertialDescendingAzimuthLaunch = Constants.PI - GetNonInertialAscendingAzimuthLaunch();
                if (_nonInertialDescendingAzimuthLaunch < 0.0)
                {
                    _nonInertialDescendingAzimuthLaunch += Constants._2PI;
                }
            }

            return _nonInertialDescendingAzimuthLaunch;
        }

        public double GetInertialAscendingAzimuthLaunch()
        {
            if (double.IsNaN(_inertialAscendingAzimuthLaunch))
            {
                double angle = TargetOrbit.Inclination();

                double latitude = LaunchSite.Planetodetic.Latitude; // In radians

                _inertialAscendingAzimuthLaunch = System.Math.Asin(System.Math.Cos(angle) / System.Math.Cos(latitude));

                if (_inertialAscendingAzimuthLaunch < 0.0)
                {
                    _inertialAscendingAzimuthLaunch += Constants._2PI;
                }
            }

            return _inertialAscendingAzimuthLaunch;
        }

        public double GetInertialDescendingAzimuthLaunch()
        {
            if (double.IsNaN(_inertialDescendingAzimuthLaunch))
            {
                _inertialDescendingAzimuthLaunch = Constants.PI - GetInertialAscendingAzimuthLaunch();
                if (_inertialDescendingAzimuthLaunch < 0.0)
                {
                    _inertialDescendingAzimuthLaunch += Constants._2PI;
                }
            }

            return _inertialDescendingAzimuthLaunch;
        }

        public double GetNonInertialInsertionVelocity()
        {
            if (double.IsNaN(_nonInertialInsertionVelocity))
            {
                double inertialInsertionVelocity = GetInertialInsertionVelocity();
                double inertialAscendingAzimuthLaunch = GetInertialAscendingAzimuthLaunch();

                double launchSiteVelocityMagnitude = LaunchSite.GetEphemeris(Time.J2000TDB, LaunchSite.CelestialBody,Frame.ICRF, Aberration.None).ToStateVector().Velocity.Magnitude();

                double vrotx = inertialInsertionVelocity * System.Math.Sin(inertialAscendingAzimuthLaunch) -
                               launchSiteVelocityMagnitude;
                double vroty = inertialInsertionVelocity * System.Math.Cos(inertialAscendingAzimuthLaunch);

                _nonInertialInsertionVelocity = System.Math.Sqrt(vrotx * vrotx + vroty * vroty);
            }

            return _nonInertialInsertionVelocity;
        }
        
        public double GetInertialInsertionVelocity()
        {
            return TargetOrbit.PerigeeVelocity();
        }

        private bool IsInboundPlaneCrossing(in Time date)
        {
            // Compute the dot product between the specific angular momentum vector of the target orbit
            // and the position vector of the launch site at the given date
            Vector3 h = TargetOrbit.ToStateVector().ToFrame(Frame.ICRF).SpecificAngularMomentum();
            Vector3 r = LaunchSite.GetEphemeris(date, LaunchSite.CelestialBody, Frame.ICRF, Aberration.None).ToStateVector().Position;
            return h * r > 0.0;
        }

        private bool IsAscending(in Time date)
        {
            // Determine if the launch will be ascending based on the dot product with the ascending node vector
            Vector3 launchSitePosition = LaunchSite.GetEphemeris(date, LaunchSite.CelestialBody, Frame.ICRF, Aberration.None).ToStateVector().Position;
            Vector3 ascendingNodeVector = TargetOrbit.ToStateVector().ToFrame(Frame.ICRF).AscendingNodeVector();
            return (launchSitePosition * ascendingNodeVector) > 0.0;
        }
    }
}

File: IO.Astrodynamics/Maneuvers/Launch.cpp
/*
 Copyright (c) 2021-2023. Sylvain Guillet (sylvain.guillet@tutamail.com)
 */
#include <Launch.h>
#include <Constants.h>
#include <InertialFrames.h>

IO::Astrodynamics::Maneuvers::Launch::Launch(const IO::Astrodynamics::Sites::LaunchSite &launchSite, const IO::Astrodynamics::Sites::Site &recoverySite, bool launchByDay,
                                   const IO::Astrodynamics::OrbitalParameters::OrbitalParameters &targetOrbit) : m_launchSite{launchSite}, m_recoverySite{recoverySite},
                                                                                                       m_launchByDay{launchByDay}, m_targetOrbit{targetOrbit}
{
    const_cast<double &>(m_inclination) = m_targetOrbit.ToStateVector().ToFrame(IO::Astrodynamics::Frames::InertialFrames::ICRF()).GetSpecificAngularMomentum().GetAngle(
            m_launchSite.GetBody()->GetBodyFixedFrame().TransformVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), IO::Astrodynamics::Math::Vector3D::VectorZ,
                                                                        IO::Astrodynamics::Time::TDB(std::chrono::duration<double>(0.0))));

    const_cast<double &>(m_deltaL) = std::asin(std::tan(m_launchSite.GetCoordinates().GetLatitude()) / std::tan(m_inclination));
}

double IO::Astrodynamics::Maneuvers::Launch::GetNonInertialAscendingAzimuthLaunch()
{
    if (std::isnan(m_nonInertialAscendingAzimuthLaunch))
    {
        double vrotx = GetInertialInsertionVelocity() * std::sin(GetInertialAscendingAzimuthLaunch()) -
                       m_launchSite.GetStateVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), IO::Astrodynamics::Time::TDB(std::chrono::duration<double>(0.0))).GetVelocity().Magnitude();
        double vroty = GetInertialInsertionVelocity() * std::cos(GetInertialAscendingAzimuthLaunch());
        m_nonInertialAscendingAzimuthLaunch = std::atan(vrotx / vroty);
        if (m_nonInertialAscendingAzimuthLaunch < 0.0)
        {
            m_nonInertialAscendingAzimuthLaunch += IO::Astrodynamics::Constants::_2PI;
        }
    }
    return m_nonInertialAscendingAzimuthLaunch;
}

double IO::Astrodynamics::Maneuvers::Launch::GetNonInertialDescendingAzimuthLaunch()
{
    if (std::isnan(m_nonInertialDescendingAzimuthLaunch))
    {
        m_nonInertialDescendingAzimuthLaunch = IO::Astrodynamics::Constants::PI - GetNonInertialAscendingAzimuthLaunch();
        if (m_nonInertialDescendingAzimuthLaunch < 0.0)
        {
            m_nonInertialDescendingAzimuthLaunch += IO::Astrodynamics::Constants::_2PI;
        }
    }
    return m_nonInertialDescendingAzimuthLaunch;
}

double IO::Astrodynamics::Maneuvers::Launch::GetInertialAscendingAzimuthLaunch()
{
    if (std::isnan(m_inertialAscendingAzimuthLaunch))
    {
        m_inertialAscendingAzimuthLaunch = std::asin(std::cos(
                m_targetOrbit.ToStateVector().ToFrame(IO::Astrodynamics::Frames::InertialFrames::ICRF()).GetSpecificAngularMomentum().GetAngle(
                        m_launchSite.GetBody()->GetBodyFixedFrame().TransformVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), IO::Astrodynamics::Math::Vector3D::VectorZ,
                                                                                    IO::Astrodynamics::Time::TDB(std::chrono::duration<double>(0.0))))) /
                                                     std::cos(m_launchSite.GetCoordinates().GetLatitude()));
        if (m_inertialAscendingAzimuthLaunch < 0.0)
        {
            m_inertialAscendingAzimuthLaunch += IO::Astrodynamics::Constants::_2PI;
        }
    }
    return m_inertialAscendingAzimuthLaunch;
}

double IO::Astrodynamics::Maneuvers::Launch::GetInertialDescendingAzimuthLaunch()
{
    if (std::isnan(m_inertialDescendingAzimuthLaunch))
    {
        m_inertialDescendingAzimuthLaunch = IO::Astrodynamics::Constants::PI - GetInertialAscendingAzimuthLaunch();
        if (m_inertialDescendingAzimuthLaunch < 0.0)
        {
            m_inertialDescendingAzimuthLaunch += IO::Astrodynamics::Constants::_2PI;
        }
    }
    return m_inertialDescendingAzimuthLaunch;
}

double IO::Astrodynamics::Maneuvers::Launch::GetNonInertialInsertionVelocity()
{
    if (std::isnan(m_nonInertialInsertionVelocity))
    {
        double vrotx = GetInertialInsertionVelocity() * std::sin(GetInertialAscendingAzimuthLaunch()) -
                       m_launchSite.GetStateVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), IO::Astrodynamics::Time::TDB(std::chrono::duration<double>(0.0))).GetVelocity().Magnitude();
        double vroty = GetInertialInsertionVelocity() * std::cos(GetInertialAscendingAzimuthLaunch());
        m_nonInertialInsertionVelocity = std::sqrt(vrotx * vrotx + vroty * vroty);
    }
    return m_nonInertialInsertionVelocity;
}

double IO::Astrodynamics::Maneuvers::Launch::GetInertialInsertionVelocity()
{
    return m_targetOrbit.GetVelocityAtPerigee();
}

std::vector<IO::Astrodynamics::Maneuvers::LaunchWindow> IO::Astrodynamics::Maneuvers::Launch::GetLaunchWindows(const IO::Astrodynamics::Time::Window<IO::Astrodynamics::Time::UTC> &searchWindow)
{
    std::vector<IO::Astrodynamics::Maneuvers::LaunchWindow> launchWindows;
    if (m_launchByDay)
    {
        //Find sunlight windows on launch site
        auto launchSiteDayWindows = m_launchSite.FindDayWindows(searchWindow, IO::Astrodynamics::Constants::OfficialTwilight);
        if (launchSiteDayWindows.empty())
        {
            throw IO::Astrodynamics::Exception::SDKException(
                    "No sunlight at launch site on this search window day : " + searchWindow.GetStartDate().ToString() + " - " + searchWindow.GetEndDate().ToString());
        }

        //Find sunlight windows on recovery site
        auto recoverySiteDayWindows = m_recoverySite.FindDayWindows(searchWindow, IO::Astrodynamics::Constants::OfficialTwilight);
        if (recoverySiteDayWindows.empty())
        {
            throw IO::Astrodynamics::Exception::SDKException(
                    "No sunlight at recovery site on this launch day : " + searchWindow.GetStartDate().ToString() + " - " + searchWindow.GetEndDate().ToString());
        }

        //Find sunlight windows on both site at same time
        std::vector<IO::Astrodynamics::Time::Window<IO::Astrodynamics::Time::UTC>> sunLightWindowsOnBothSites;
        for (auto &&launchSiteWindow: launchSiteDayWindows)
        {
            for (auto &&recoverySiteWindow: recoverySiteDayWindows)
            {
                if (launchSiteWindow.Intersects(recoverySiteWindow))
                {
                    auto intersection = launchSiteWindow.GetIntersection(recoverySiteWindow);
                    sunLightWindowsOnBothSites.emplace_back(intersection.GetStartDate(), intersection.GetEndDate());
                }
            }
        }

        if (sunLightWindowsOnBothSites.empty())
        {
            throw IO::Astrodynamics::Exception::SDKException("No sun light at same time on both Sites");
        }

        //Search an orbital plane alignment with launch site during sunlight window on both Sites
        for (auto &&sunlightWindow: sunLightWindowsOnBothSites)
        {
            auto res = FindLaunchWindows(sunlightWindow);
            for (auto &lw: res)
            {
                launchWindows.push_back(lw);
            }
        }
    } else
    {
        auto res = FindLaunchWindows(searchWindow);
        for (auto &lw: res)
        {
            launchWindows.push_back(lw);
        }
    }

    return launchWindows;
}

std::vector<IO::Astrodynamics::Maneuvers::LaunchWindow> IO::Astrodynamics::Maneuvers::Launch::FindLaunchWindows(const IO::Astrodynamics::Time::Window<IO::Astrodynamics::Time::UTC> &windowToSearch)
{
    //Initialize start date
    auto date = windowToSearch.GetStartDate();

    //Define initial step
    auto step = windowToSearch.GetLength() * 0.5;

    //Define crossing plane inbound status
    bool isInboundPlaneCrossing = m_targetOrbit.ToStateVector().ToFrame(IO::Astrodynamics::Frames::InertialFrames::ICRF()).GetSpecificAngularMomentum().DotProduct(
            m_launchSite.GetStateVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), date.ToTDB()).GetPosition()) > 0.0;

    std::vector<LaunchWindow> launchWindows;

    //Works until it reaches the end of search window
    while (date <= windowToSearch.GetEndDate())
    {
        date = date + step;

        //Compute crossing plane inbound status
        bool isInbound = m_targetOrbit.ToStateVector().ToFrame(IO::Astrodynamics::Frames::InertialFrames::ICRF()).GetSpecificAngularMomentum().DotProduct(
                m_launchSite.GetStateVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), date.ToTDB()).GetPosition()) > 0.0;

        //If inbound status has changed, we passed through orbital plane
        if (isInbound != isInboundPlaneCrossing)
        {
            isInboundPlaneCrossing = isInbound;
            //We reduce step size and reverse search direction
            step = step * -0.5;
        }

        //If a sufficent accuracy is reached
        if (std::abs(step.GetSeconds().count()) < 1.0)
        {
            //Fill in the first launch window

            double inertialAzimuthLaunch{std::numeric_limits<double>::quiet_NaN()};
            double nonInertialAzimuthLaunch{std::numeric_limits<double>::quiet_NaN()};

            //Set velocities
            double inertialInsertionVelocity = GetInertialInsertionVelocity();
            double nonInertialInsertionVelocity{GetNonInertialInsertionVelocity()};

            //Define if launch will be northly or southerly and set azimuths
            bool isAscending{};
            if (m_launchSite.GetStateVector(IO::Astrodynamics::Frames::InertialFrames::ICRF(), date.ToTDB()).GetPosition().DotProduct(
                    m_targetOrbit.ToStateVector().ToFrame(IO::Astrodynamics::Frames::InertialFrames::ICRF()).GetAscendingNodeVector()) > 0.0)
            {
                inertialAzimuthLaunch = GetInertialAscendingAzimuthLaunch();
                nonInertialAzimuthLaunch = GetNonInertialAscendingAzimuthLaunch();
                isAscending = true;
            } else
            {
                inertialAzimuthLaunch = GetInertialDescendingAzimuthLaunch();
                nonInertialAzimuthLaunch = GetNonInertialDescendingAzimuthLaunch();
            }

            //We add the first launch window in the collection
            auto window = IO::Astrodynamics::Time::Window<IO::Astrodynamics::Time::UTC>(date, date);
            launchWindows.emplace_back(m_launchSite, window, inertialAzimuthLaunch, nonInertialAzimuthLaunch, inertialInsertionVelocity,
                                                                     nonInertialInsertionVelocity);

            //We compute following launch windows
            auto remainingTime = windowToSearch.GetEndDate() - date;

            //sideral rotation period
            auto sideralRotation = m_launchSite.GetBody()->GetSideralRotationPeriod(date.ToTDB());
            auto halfSideralRotation = sideralRotation * 0.5;

            //nb remaining launch windows
            int nbLaunchWindows = remainingTime.GetSeconds().count() / halfSideralRotation.GetSeconds().count();

            for (int i = 0; i < nbLaunchWindows; i++)
            {
                //If previous is ascending next will be descending
                isAscending = !isAscending;

                auto deltalDuration = m_deltaL * 2.0 / m_launchSite.GetBody()->GetAngularVelocity(date.ToTDB());

                if (!isAscending)
                {
                    deltalDuration *= -1.0;
                }



                //We increment next launch date
                date = date + halfSideralRotation + IO::Astrodynamics::Time::TimeSpan(std::chrono::duration<double>(deltalDuration));

                //build next window
                auto nextWindow = IO::Astrodynamics::Time::Window<IO::Astrodynamics::Time::UTC>(date, date);

                if (isAscending)
                {
                    launchWindows.emplace_back(m_launchSite, nextWindow, GetInertialAscendingAzimuthLaunch(), GetNonInertialAscendingAzimuthLaunch(),
                                                                             inertialInsertionVelocity, nonInertialInsertionVelocity);
                } else
                {
                    launchWindows.emplace_back(m_launchSite, nextWindow, GetInertialDescendingAzimuthLaunch(), GetNonInertialDescendingAzimuthLaunch(),
                                                             inertialInsertionVelocity, nonInertialInsertionVelocity);
                }
            }
            break;
        }
    }

    return launchWindows;
}
